src/JustEat.StatsD/Buffered/BufferBasedStatsDPublisher.cs

```diff
@@ -0,0 +1,184 @@
+using System;
+
+namespace JustEat.StatsD.Buffered
+{
+    internal sealed class BufferBasedStatsDPublisher : IStatsDPublisher
+    {
+        private const double DefaultSampleRate = 1.0;
+        private const int SafeUdpPacketSize = 512;
+
+        [ThreadStatic]
+        private static byte[] _buffer;
+        private static byte[] Buffer => _buffer ?? (_buffer = new byte[SafeUdpPacketSize]);
+
+        [ThreadStatic]
+        private static Random _random;
+        private static Random Random => _random ?? (_random = new Random());
+
+        private readonly StatsDUtf8Formatter _formatter;
+        private readonly IStatsDBufferedTransport _transport;
+        private readonly Func<Exception, bool> _onError;
+
+        public BufferBasedStatsDPublisher(StatsDConfiguration configuration, IStatsDBufferedTransport transport)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            _onError = configuration.OnError;
+            _transport = transport;
+            _formatter = new StatsDUtf8Formatter(configuration.Prefix);
+        }
+
+        public void Increment(string bucket)
+        {
+            Increment(1, bucket);
+        }
+
+        public void Increment(long value, string bucket)
+        {
+            Increment(value, DefaultSampleRate, bucket);
+        }
+
+        public void Increment(long value, double sampleRate, string bucket)
+        {
+            var msg = StatsDMessage.Counter(value, bucket);
+            SendMessage(sampleRate, msg);
+        }
+
+        public void Increment(long value, double sampleRate, params string[] buckets)
+        {
+            if (buckets == null)
```

Comment:
Could also shortcut if `buckets.Length == 0` to remove the need to allocate the enumerator.